src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/QualityGuidelines/AvoidMultipleEnumerations/AvoidMultipleEnumerationsHelpers.cs

```diff
@@ -0,0 +1,483 @@
+
+        /// <summary>
+        /// Check if <param name="operation"/> is an argument that passed into a linq chain. (like Select, Where etc.)
+        /// </summary>
+        private static bool IsOperationIsArgumentOfLinqChainInvocation(
+            IOperation operation, WellKnownSymbolsInfo wellKnownSymbolsInfo, out bool enumerateArgument)
+        {
+            if (operation is IArgumentOperation { Parent: IInvocationOperation invocationOperation } argumentOperation)
+            {
+                return IsLinqChainInvocation(invocationOperation, argumentOperation, wellKnownSymbolsInfo, out enumerateArgument);
+            }
+
+            enumerateArgument = false;
+            return false;
+        }
+
+        /// <summary>
+        /// Check if <param name="argumentOperationToCheck"/> is passed as a deferred executing argument into <param name="invocationOperation"/>.
+        /// </summary>
+        public static bool IsLinqChainInvocation(
+            IInvocationOperation invocationOperation,
+            IArgumentOperation argumentOperationToCheck,
+            WellKnownSymbolsInfo wellKnownSymbolsInfo,
+            out bool enumerateArgument)
+        {
+            enumerateArgument = false;
+            RoslynDebug.Assert(invocationOperation.Arguments.Contains(argumentOperationToCheck));
+            if (!IsDeferredType(argumentOperationToCheck.Value.Type?.OriginalDefinition, wellKnownSymbolsInfo.AdditionalDeferredTypes))
+            {
+                return false;
+            }
+
+            var targetMethod = invocationOperation.TargetMethod;
+            // For C#, extension method is used as an ordinary static method.
+            // For VB, ex: a.Concat(b)
+            // 'b' is an argument to 'Concat', which is a reduced method.
+            var reducedFromMethod = targetMethod.ReducedFrom ?? targetMethod;
+            var originalMethod = reducedFromMethod.OriginalDefinition;
+            if (!IsDeferredType(targetMethod.ReturnType.OriginalDefinition, wellKnownSymbolsInfo.AdditionalDeferredTypes))
+            {
+                return false;
+            }
+
+            // For VB, ex: a.Concat(b)
+            // 'b' is in fact the first argument to 'Concat', because the extension method in VB is reduced.
+            var argumentMappingParameter = targetMethod.MethodKind == MethodKind.ReducedExtension
+                ? GetReducedFromParameter(targetMethod, argumentOperationToCheck.Parameter)
+                : argumentOperationToCheck.Parameter;
+
+            if (wellKnownSymbolsInfo.LinqChainMethods.Contains(originalMethod)
+                && originalMethod.Parameters.Any(
+                    methodParameter => IsDeferredType(methodParameter.Type?.OriginalDefinition, wellKnownSymbolsInfo.AdditionalDeferredTypes) && methodParameter.Equals(argumentMappingParameter.OriginalDefinition)))
+            {
+                // All well-known linq chain methods under Linq namespace won't enumerate the argument.
```

Comment:
NIT: `if it won't enumerate the argument` then why returning true?

Reply:
It is saying that for method under linq namespace, like 'Select', 'Where'...
Call `a.Select(i => i + 1)` won't enumerate `a` 
There is an `out bool enumerateArgument` in this method to indicate this.
I shall update this comment to make it more clear

Reply:
Ah the comment was for that `enumerateArgument` not for the returning value `true`